### PR TITLE
add error handling to bcadmin and extend 'key' and 'dump' command

### PR DESCRIPTION
**Pull request by ineiti**
byzcoin/bcadmin was missing a lot of error handling, which has been updated.
The 'key' command got '--print' to show the private part of the key.
The 'dump' command got '--verbose' to print some details of the printed instances.

### CODE REVIEW

File: byzcoin/bcadmin/main.go [hunk 1]
```diff
@@ -477,20 +493,24 @@ func link(c *cli.Context) error {
 		}
 		ad := &darc.Darc{}
 		adPub := cothority.Suite.Point()
-		if adIDStr := c.String("darc"); len(adIDStr) == 69 {
-			adID, err := hex.DecodeString(adIDStr[5:])
+		// Accept both plain-darcs, as well as "darc:...." darcs
```
Comment (jeffallen) [inline byzcoin/bcadmin/main.go]: getDarcByString is already doing "strings.HasPrefix("darc:")" and then trimming it. Is there any way we can keep this in one place?

File: byzcoin/bcadmin/main.go [hunk 2]
```diff
@@ -477,20 +493,24 @@ func link(c *cli.Context) error {
...
-			if pubStr := c.String("pub"); len(pubStr) == 72 {
-				adPubBuf, err = hex.DecodeString(pubStr[8:])
-				if len(adPubBuf) != 32 || err != nil {
-					return errors.New("please give valid admin public key in hex: " + err.Error())
-				}
-			} else {
-				return errors.New("please give a key in the following format: ed25519:public_key_in_hex")
+			adPubStr := c.String("adminpub")
+			// Accept both plain-public keys and "ed25519:..." keys
+			if (len(adPubStr) != 64 && len(adPubStr) != 72) || err != nil {
```
Comment (jeffallen) [inline byzcoin/bcadmin/main.go]: Am really unconvinced on this length-based hacking around. If we are going to hack around on the input, strings.HasPrefix is much more self-documenting.

File: byzcoin/bcadmin/main.go [hunk 3]
```diff
@@ -399,8 +409,14 @@ func create(c *cli.Context) error {
 		return err
 	}
 
-	fmt.Fprintf(c.App.Writer, "Created ByzCoin with ID %x.\n", cfg.ByzCoinID)
-	fmt.Fprintf(c.App.Writer, "export BC=\"%v\"\n", fn)
+	_, err = fmt.Fprintf(c.App.Writer, "Created ByzCoin with ID %x.\n", cfg.ByzCoinID)
+	if err != nil {
+		return err
```
Comment (jeffallen) [inline byzcoin/bcadmin/main.go]: What problem are you trying to solve here? Yes, we know that Printf can fail, but was it really failing? It just seems a little excessive, and wondering how you decided these needed to be checked...

I'm not saying to remove these checks, just curious...

File: byzcoin/proto.go [hunk 4]
```diff
@@ -57,7 +57,7 @@ type CreateGenesisBlock struct {
 	Version Version
 	// Roster defines which nodes participate in the skipchain.
 	Roster onet.Roster
-	// GenesisDarc defines who is allowed to write to this skipchain.
+	// AdminDarc defines who is allowed to write to this skipchain.
```
Comment (jeffallen) [inline byzcoin/proto.go]: What?

File: byzcoin/service.go [hunk 5]
```diff
@@ -1875,6 +1875,7 @@ func (s *Service) processOneTx(sst *stagingStateTrie, tx ClientTransaction) (Sta
 				}
 				return nil, nil, fmt.Errorf("%s: contract %s %s", s.ServerIdentity(), contractID, reason)
 			}
+			log.Lvlf2("StateChange %s for id %x - contract: %s", sc.StateAction, sc.InstanceID, sc.ContractID)
```
Comment (jeffallen) [inline byzcoin/service.go]: Are you sure level 2 is correct here? Seems like a 3 thing?

File: byzcoin/bcadmin/main.go (hunk 3 above)
Comment (ineiti) [reply byzcoin/bcadmin/main.go]: Just following IntelliJ's yellow lines...

File: byzcoin/proto.go (hunk 4 above)
Comment (ineiti) [reply byzcoin/proto.go]: When I wanted to connect to the DEDIS-ByzCoin, you created a DARC for me, which is now _my_ admin darc. So the darc stored in the config is not necessarily the genesis-darc, but only any admin darc that is allowed to do things in this ByzCoin instance.

File: byzcoin/service.go (hunk 5 above)
Comment (ineiti) [reply byzcoin/service.go]: I would say that until we have a lot of transactions, it's really a useful debugging output. If we start having multi-transaction blocks regularly, we can change it to `Lvl3`. What do you think?

File: byzcoin/bcadmin/main.go (hunk 1 above)
Comment (ineiti) [reply byzcoin/bcadmin/main.go]: OK, split `getDarcByString` to a string-to-id and a `getDarc` part. Also added `stringToEd25519Buf`.

File: byzcoin/proto.go (hunk 4 above)
Comment (jeffallen) [reply byzcoin/proto.go]: Right, but the field is still called GenesisDarc, so the documentation string needs to start with, at least, a capital G...

File: byzcoin/service.go (hunk 5 above)
Comment (jeffallen) [reply byzcoin/service.go]: ok, makes sense.

File: byzcoin/proto.go (hunk 4 above)
Comment (ineiti) [reply byzcoin/proto.go]: Thanks - now I got it. Yes, it's IntelliJ over-zealously replacing `GenesisDarc` with `AdminDarc`, and me not reading the code on Friday afternoon...

**Pull request merged**